src/main/java/com/google/devtools/build/lib/skyframe/RepoFileFunction.java

```diff
@@ -87,29 +89,23 @@ public SkyValue compute(SkyKey skyKey, Environment env)
     }
     if (!repoFileValue.exists()) {
       // It's okay to not have a REPO.bazel file.
-      return RepoFileValue.of(PackageArgs.EMPTY);
+      return RepoFileValue.of(ImmutableMap.of(), ImmutableList.of());
     }
 
     // Now we can actually evaluate the file.
     StarlarkSemantics starlarkSemantics = PrecomputedValue.STARLARK_SEMANTICS.get(env);
-    RepositoryMappingValue repoMapping =
-        (RepositoryMappingValue) env.getValue(RepositoryMappingValue.key(repoName));
-    RepositoryMappingValue mainRepoMapping =
-        (RepositoryMappingValue) env.getValue(RepositoryMappingValue.key(RepositoryName.MAIN));
     if (env.valuesMissing()) {
       return null;
     }
-    StarlarkFile repoFile = readAndParseRepoFile(repoFilePath.asPath(), env);
-    PackageArgs packageArgs =
-        evalRepoFile(
-            repoFile,
-            repoName,
-            repoMapping.getRepositoryMapping(),
-            mainRepoMapping.getRepositoryMapping(),
-            starlarkSemantics,
-            env.getListener());
 
-    return RepoFileValue.of(packageArgs);
+    StarlarkFile repoFile = readAndParseRepoFile(repoFilePath.asPath(), env);
+    return evalRepoFile(
+        repoFile,
+        repoName,
+        RepositoryMapping.ALWAYS_FALLBACK,
```

Comment:
actually, how is this split helping things? It looks like we end up always using `null` for the main repo mapping, which ends up being the same as cutting edge 8. So the split is actually not even doing anything?

(I expected the split to be done in a way that runs the REPO.bazel file twice: the first time without a main repo mapping just to get the ignore list, and the second time _with_ the main repo mapping to get the repo's PackageArgs. If we're not running the REPO.bazel file twice, there doesn't seem to be any point in splitting.)

Reply:
I see that, but my point is that netiher the first nor the second function is actually using the **main** repo mapping in this PR. This means we wouldn't have needed the split in the first place (IIUC the cycle was caused by the _main_ repo mapping, not the mapping of the repo where REPO.bazel is being processed). IOW, why can't IgnoredPrefixes just use the value of the second function?

Reply:
There were two cycles, the one you linked above and apparently also https://github.com/bazelbuild/bazel/pull/24022#issuecomment-2419309453. The latter goes through the current repo's repo mapping, which may be the main repo mapping.

Reply:
oh, I missed that other PR... annoying, indeed.

Could we avoid this whole dance by having a version of package lookup that doesn't care about ignored packages? Might not be the best idea... let me think about it more.

Reply:
FWIW, I think the current approach is better than  two versions of the package lookup because that one goes much deeper into the internals of Blaze and also much more confusing because it introduces another way to lookup packages that's almost, but not quite, the same as the "regular" one.

The reason why I like the two-phase approach is that it's conceptually sound: one phase parses `REPO.bazel`, then another digests its labels. Yes, it adds some complications, but it's the the cleanest one, at least architecturally.  

Once `WORKSPACE` is gone, we can delete this again (if I understand it correctly). Is that enough consolation?

Reply:
we actually can't delete this even when WORKSPACE is gone, because computing repo mappings in Bzlmod can also require package lookup first (e.g. `archive_override` with a patch). But I agree that the current approach is the best we can do.